<launch>
    <!-- 打开webots -->
    <include file="$(find webots_demo)/launch/webots.launch" />
    <arg name="model" />
    <arg name="gui" default="false" />
    <!-- 载入机器人模型 -->
    <param name="robot_description" textfile="$(find webots_demo)/urdf/demo.urdf" />
    <param name="use_gui" value="$(arg gui)" />
    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" />
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />
    <node name="robot_broadcaster" pkg="webots_demo" type="robot_broadcaster" />
    <!-- 打开rviz -->
    <node name="rviz" pkg="rviz" type="rviz"  args="-d $(find webots_demo)/rviz/demo.rviz"/>
    
    <!-- Gmapping建图算法，接收传感器数据建立地图 -->
    <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" output="screen" clear_params="true">
    <param name="odom_frame" value="odom" />
    <param name="base_frame" value="base_link" />
    <param name="map_update_interval" value="0.75" />
    <!-- Set maxUrange < actual maximum range of the Laser -->
    <param name="maxRange" value="4.5" />
    <param name="maxUrange" value="3.5" />
    <param name="sigma" value="0.05" />
    <param name="kernelSize" value="1" />
    <param name="lstep" value="0.05" />
    <param name="astep" value="0.05" />
    <param name="iterations" value="5" />
    <param name="lsigma" value="0.075" />
    <param name="ogain" value="3.0" />
    <param name="lskip" value="0" />
    <param name="srr" value="0.01" />
    <param name="srt" value="0.02" />
    <param name="str" value="0.01" />
    <param name="stt" value="0.02" />
    <param name="linearUpdate" value="0.5" />
    <param name="angularUpdate" value="0.157" />
    <param name="temporalUpdate" value="-1.0" />
    <param name="resampleThreshold" value="0.5" />
    <param name="particles" value="50" />

    <param name="xmin" value="-1.0" />
    <param name="ymin" value="-1.0" />
    <param name="xmax" value="1.0" />
    <param name="ymax" value="1.0" />

    <param name="delta" value="0.05" />
    <param name="llsamplerange" value="0.01" />
    <param name="llsamplestep" value="0.01" />
    <param name="lasamplerange" value="0.005" />
    <param name="lasamplestep" value="0.005" />
    <remap from="scan" to="/robot/Sick_LMS_291/laser_scan/layer0" />
  </node>
</launch>
<!-- rosrun gmapping slam_gmapping scan:=/volcano/Sick_LMS_291/laser_scan/layer0 _xmax:=30 _xmin:=-30 _ymax:=30 _ymin:=-30 _delta:=0.2' -->